packages/editor/src/components/document-bar/index.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { __, isRTL, sprintf } from '@wordpress/i18n';
+import { useSelect, useDispatch } from '@wordpress/data';
+import {
+	Button,
+	__experimentalText as Text,
+	__experimentalHStack as HStack,
+} from '@wordpress/components';
+import { BlockIcon } from '@wordpress/block-editor';
+import {
+	chevronLeftSmall,
+	chevronRightSmall,
+	page as pageIcon,
+	navigation as navigationIcon,
+	symbol,
+} from '@wordpress/icons';
+import { displayShortcut } from '@wordpress/keycodes';
+import { useEntityRecord } from '@wordpress/core-data';
+import { store as commandsStore } from '@wordpress/commands';
+import { useState, useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as editorStore } from '../../store';
+
+const typeLabels = {
+	// translators: 1: Pattern title.
+	wp_pattern: __( 'Editing pattern: %s' ),
+	// translators: 1: Navigation menu title.
+	wp_navigation: __( 'Editing navigation menu: %s' ),
+	// translators: 1: Template title.
+	wp_template: __( 'Editing template: %s' ),
+	// translators: 1: Template part title.
+	wp_template_part: __( 'Editing template part: %s' ),
+};
+
+const icons = {
+	wp_block: symbol,
+	wp_navigation: navigationIcon,
+};
+
+export default function DocumentBar() {
+	const { isEditingTemplate, templateId, postType, postId } = useSelect(
+		( select ) => {
+			const {
+				getRenderingMode,
+				getCurrentTemplateId,
+				getCurrentPostId,
+				getCurrentPostType,
+			} = select( editorStore );
+			const _templateId = getCurrentTemplateId();
+			return {
+				isEditingTemplate:
+					!! _templateId && getRenderingMode() === 'template-only',
+				templateId: _templateId,
+				postType: getCurrentPostType(),
+				postId: getCurrentPostId(),
+			};
+		},
+		[]
+	);
+	const { getEditorSettings } = useSelect( editorStore );
+	const { setRenderingMode } = useDispatch( editorStore );
+
+	return (
+		<BaseDocumentActions
+			postType={ isEditingTemplate ? 'wp_template' : postType }
+			postId={ isEditingTemplate ? templateId : postId }
+			onBack={
+				isEditingTemplate
+					? () =>
+							setRenderingMode(
+								getEditorSettings().defaultRenderingMode
+							)
```

Comment:
Why is this limited only to template types? Navigation Menu posts (`wp_navigation`) are also edited in this context and it seems like "back" default rendering mode would be useful here also. I'm just checking whether this was intentionally locked down to templates only.

Reply:
Oh I see, this is about rendering modes.

Reply:
How would you feel about the default "back" action always be to go "back" in the `@wordpress/router` history? 

eg. `history.back()`

Design would like this "Back" button to become the default way to go back out of the template editing mode and I need to implement this in https://github.com/WordPress/gutenberg/pull/55657.

Reply:
The post editor doesn't use the router, so this would break.

Reply:
I also feel like "back" using "history.back" is a bit redundant with the browser's back button. This "back" is more a  "exist template mode" button.

Reply:
I understand that perspective. 

Multiple user testing sessions I've been involved with recently have indicated that users become quite confused about how to exit this mode however. The browser back button isn't as obvious as perhaps we expect.

I know @jasmussen is keen to see this become the generic "Back" button.

The specifics of how we go "back" need to be worked out. As you say, router might not the right path.

Reply:
What happens when you click the back button I'll defer to you for best practices, though I'd tend to agree for the issue Rich highlighted in another post, history.back may fall short in some cases. 

So to clarify generic back button, I mean it mostly in that the document title where a back button sits on the left is an excellent interface to unify on as the position of said button. Not just because it's in a consistent place next to related UI, it also has a basic breadcrumb-like behavior which is useful if you go into multiple levels of focus mode, and finally we have mockups that make this work on mobile, which we don't have with the back button in the canvas.

Reply:
Thanks for going over this again Joen. So hopefully we can agree that UI-wise we'd like the "Back" in the top toolbar to become the default UI control for that action?

@youknowriad Technically, could we add a `referrer` prop to the URL which we can then consume to determine the best location for "back". If none is provided we optimise for "go back to the main site editor view"?

Obviously there would need to be an exception for situations like what is currently handled in this PR where "Back" means "Exit template editing mode" but that's simple enough. What do you think?

Reply:
To be honest, it's not satisfying for me. I think we're breaking a big conceptual element in order to satisfy a small affordance. 

Think of the whole editor or frame as `<Editor postType={ currentPostType } postId={ currentPostId } />`. So the whole editor is just a "reusable component" that you can use in any context (post editor, site editor, elsewhere...).

Also the editor understands "template modes" so you can do `<Editor postType={postType} postId={postId} templateId={templateId} />` and all the switching of modes (from post+template to template only and back...) is all internal. So in that sense, a "back" button within the editor to switch between these modes make sense.

Now, think of focus modes: focus modes mean switching the actual "postType" and "postId" passed to the editor, so they are things that are external to the editor. It's a shame to have to update the editor internally for this affordance. Also and more imporantly, it might not make sense in all these "editors" and "frames". For me, "focus mode" means navigating to another entity in the site editor, you actually switch a lot of things, not just the UI of the editor, you switch the active page in the sidebar of the site editor, you switch the URL..., you're not within the same editor anymore, it's another one. So it feels weird to have a back button **within** the editor. For me it's a cross page affordance and in that sense is better placed **outside** the editor.

Now, that is my opinion and my strong opinion here is that we shouldn't break the Editor abstraction just for this. That said, If you all feel very strongly about this "back" button, we can try to find a compromise and have a "setting" in the settings passed to `EditorProvider` that contains the "back referrer" or something like that. Maybe even make it a private setting (private API)

But I persist, that I think it's wrong personally :)

Reply:
Can we agree that when users enter focus mode for a template part they need to be given an easy way back to the template they were editing?

Reply:
> better placed outside the editor

Do you mean literally? The idea of a notice bar at the bottom of the viewport has come up a few times but never really gained much traction.

<img width="1486" alt="Screenshot 2023-12-07 at 11 00 51" src="https://github.com/WordPress/gutenberg/assets/846565/d4cda2a4-e1de-426e-975d-6c877bd08307">

Worth noting theres a 'Back to x' suggestion in the command palette when you click the document bar, so I'd agree with Joen – the current placement feels natural in terms of UX.

Reply:
After discussing more with @jasmussen It seems the main point of friction is that:

 - we have "focus mode" for template part, navigation block and things like that where in fact we actually navigate to another page of the site editor.
 - we have "template mode" in the site editor where we show a template, but we don't call that "focus mode" and we don't navigate to another page of the site editor.

If we are to treat the "back" button as similar for both these cases, they should first work similarly. In other words, maybe the "template mode" should be removed and instead we should navigate to the "template page" of the site editor instead.

Also, if we do that, and since the "template mode" is also an affordance of the "post editor", we should make sure to also add some kind of "template page" to the post editor too (maybe no url navigation there or anything but at least it should be something in the "state" of the post editor and the post editor should be able to switch).